mapproxy/grid/tile_grid.py

```diff
@@ -83,6 +83,87 @@ def tile_grid(srs=None, bbox=None, bbox_srs=None, tile_size=(256, 256),
                     origin=origin, name=name)
 
 
+class UnsupportedException(Exception):
+    def __init__(self, arg):
+        Exception.__init__(self, arg)
+
+
+def tile_grid_from_ogc_tile_matrix_set(ogc_tile_matrix_set):
+    """ Return a TileGrid instance from the JSON representation of a
+        OGC Two Dimensional Tile Matrix Set (https://docs.ogc.org/is/17-083r4/17-083r4.html)
+    """
+
+    srs = ogc_tile_matrix_set["crs"]
+    srs = SRS(ogc_crs_url_to_auth_code(srs))
+    tileMatrices = ogc_tile_matrix_set["tileMatrices"]
+    tileWidth = None
+    tileHeight = None
+    cornerOfOrigin = None
+    origin = None
+    bbox = None
+    res = []
+    grid_sizes = []
+    for tileMatrix in tileMatrices:
+        if "variableMatrixWidths" in tileMatrix:
+            raise UnsupportedException("Tile matrix with variableMatrixWidths are not supported")
+
+        cellSize = tileMatrix["cellSize"]
+        matrixWidth = tileMatrix["matrixWidth"]
+        matrixHeight = tileMatrix["matrixHeight"]
+
+        if tileWidth is None:
+            tileWidth = tileMatrix["tileWidth"]
+            tileHeight = tileMatrix["tileHeight"]
+            pointOfOrigin = tileMatrix["pointOfOrigin"]
+            if srs.is_axis_order_ne:
+                x = pointOfOrigin[1]
+                y = pointOfOrigin[0]
+            else:
+                x = pointOfOrigin[0]
+                y = pointOfOrigin[1]
+
+            cornerOfOrigin = tileMatrix["cornerOfOrigin"]
+            if cornerOfOrigin == "topLeft":
+                origin = 'ul'
+                bbox = [x,
+                        y - tileHeight * matrixHeight * cellSize,
+                        x + tileWidth * matrixWidth * cellSize,
+                        y]
+            elif cornerOfOrigin == "bottomLeft":
+                origin = 'll'
+                bbox = [x, y,
+                        x + tileWidth * matrixWidth * cellSize,
+                        y + tileHeight * matrixHeight * cellSize]
+            else:
+                raise UnsupportedException(f"Invalid cornerOfOrigin={cornerOfOrigin}")
```

Comment:
It would be nice if this could be refactored to its own method for better testing